runtime/Dart/lib/src/token.dart

```diff
@@ -144,16 +142,18 @@ class CommonToken extends WritableToken {
   ///
   /// @param type The token type.
   /// @param text The text of the token.
-  CommonToken(this.type,
-      {this.source = EMPTY_SOURCE,
-      this.channel = Token.DEFAULT_CHANNEL,
-      this.startIndex,
-      this.stopIndex,
-      text}) {
+  CommonToken(
+    this.type, {
+    this.source = EMPTY_SOURCE,
+    this.channel = Token.DEFAULT_CHANNEL,
+    this.startIndex = -1,
```

Comment:
i defaulted these to -1 to avoid making them nullable, does it make sense?

Reply:
Would it be better to make it a nullable value to indicate a potential invalid state?

Reply:
I don't think nullable makes sense here, as it will make the `startindex` in `Token` interface nullable.
In Java, `startIndex` has the type `int` which is not nullable with a default value of 0. I suggest doing the same.

https://github.com/antlr/antlr4/blob/b5648f405b746d87ec559132ad1939fb6d402053/runtime/Java/src/org/antlr/v4/runtime/CommonToken.java#L74

Reply:
I went with -1 based on the documentation on https://github.com/antlr/antlr4/blob/master/runtime/Dart/lib/src/token.dart#L76

But I guess that unimplemented and not-set might mean different things, so I'll change it to zero.

Reply:
@lingyv-li Help me understand why the interface should not be nullable if we're defining a sort of "null state" anyway? To me it seems like we're using -1 as a fake null, which might obfuscate the code if there's not a good reason for it.

Reply:
That's a good callout. I didn't pay attention to the documentation so I see how nullable may be a better choice here now.
However I still think we should keep it non-nullable, mainly because:
The Dart target implementation is heavily based on Java target, where if you look at the code you can almost always find 1:1 correspondence. This is because the Java implementation is performant and reliable, and for future maintainence it would be easiest if we keep this correspondence. I see that Dart and Java are different languages after all, and at some point we have to diverge. But in this case I don't think we have a good enough reason to do that.

Reply:
That's fair, I didn't think about maintenance -- divergent code would make future changes more difficult, I agree there's more value in that.

Reply:
Done, defaulted to zero 👍🏽